topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -133,72 +136,102 @@ public void init(Vertx vertx, Context context) {
+            .compose(ignored -> {
+                try {
+                    shutdownExecutor.shutdown();
+                    shutdownExecutor.awaitTermination(10, TimeUnit.SECONDS);
+                } catch (InterruptedException e) {
+                    if (!shutdownExecutor.isTerminated()) {
+                        shutdownExecutor.shutdownNow();
                     }
                 }
-            };
-            longHandler.handle(null);
-
-            promise.future().compose(ignored -> {
-                if (service != null) {
-                    service.stop();
-                }
-                return Future.succeededFuture();
+                LOGGER.info("Stopped");
+                return Future.future(handle -> stop.complete());
             });
+    }
 
-            promise.future().compose(ignored -> {
+    private Future<Void> closeKubernetesWatch() {
+        LOGGER.debug("Closing Kubernetes watch");
+        if (topicWatch != null) {
+            topicWatch.close();
+        }
+        return Future.succeededFuture();
+    }
 
-                LOGGER.debug("Disconnecting from zookeeper {}", zk);
-                zk.disconnect(zkResult -> {
-                    if (zkResult.failed()) {
-                        LOGGER.warn("Error disconnecting from zookeeper: {}", String.valueOf(zkResult.cause()));
-                    }
-                    long timeoutMs = Math.max(1, deadline - System.currentTimeMillis());
-                    LOGGER.debug("Closing AdminClient {} with timeout {}ms", adminClient, timeoutMs);
-                    try {
-                        adminClient.close(Duration.ofMillis(timeoutMs));
-                        HttpServer healthServer = this.healthServer;
-                        if (healthServer != null) {
-                            healthServer.close();
-                        }
-                    } catch (TimeoutException e) {
-                        LOGGER.warn("Timeout while closing AdminClient with timeout {}ms", timeoutMs, e);
-                    } finally {
-                        LOGGER.info("Stopped");
-                        blockingResult.complete();
+    private Future<Void> stopZooKeeperWatcher() {
+        LOGGER.debug("Stopping ZooKeeper watcher");
+        if (topicsWatcher != null) {
+            topicsWatcher.stop();
+        }
+        return Future.succeededFuture();
+    }
+
+    private Future<Void> stopTopicStoreService() {
+        LOGGER.debug("Stopping topic store service");
+        if (service != null) {
+            service.stop();
+        }
+        return Future.succeededFuture();
+    }
+
+    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
+    private Future<Void> disconnectFromZk(ExecutorService shutdownExecutor) {
+        LOGGER.debug("Disconnecting from ZooKeeper");
+        Promise<Void> promise = Promise.promise();
+        if (zk != null) {
+            shutdownExecutor.submit(() -> {
+                zk.disconnect(result -> {
```

Comment:
We need to be very careful (and should document) on which thread the returned futures complete. In this case that depends on `ZkImpl`, and to be entirely honest I'm not quite sure which thread that will be, because we're calling disconnect from a thread that's not vertx-managed, so I don't know if it's able to recover the/a context thread.